### ISSUE TITLE
[css-lists] Should option/optgroup be able to set counters?

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#4004]
Should one be able to use 'counter-increment'/etc on `<option>`/`<optgroup>` (and other replaced-element form-control descendants)?

I think a strict reading of the spec says no - since `<select>` is replaced, its children don't generate CSS boxes, and things that don't generate CSS boxes can't interact with counters.

However, that text was written with `display:none` in mind, not necessarily replaced element descendants.

Here's an example: <https://codepen.io/kizu/pen/PEJgpr>. Apparently this works in Firefox, and pre-chromium Edge, but it does not work in WebKit or Chrome.

**Comment by fantasai** [comment]
Hmm, yeah. I think that should work. So CSS boxes or other representation in a rendering tree that's capable of holding CSS properties???

**Comment by kizu** [comment]
Small note: it works in Chrome with `multiple` attribute :) https://codepen.io/kizu/pen/EowqxX

(Also, [a link to the chromium issue](https://bugs.chromium.org/p/chromium/issues/detail?id=799465#c7))

I'm not sure if I want to push this as something that should work everywhere, but there is no other way right now to use the value of selects in any other way in CSS, so…

Other case tests for which that I remember doing and which had differences in various browsers (but I don't think I created issues for this) — `counter-increment` inside inline SVG. I imagine there could be other _interesting_ cases, including maybe shadow DOM and so on :D

**Comment by tabatkins** [comment]
Yeah, `svg` is also a replaced element, and its descendants don't generate CSS boxes either...

**Comment by tabatkins** [comment]
Based on testing with <http://software.hixie.ch/utilities/js/live-dom-viewer/saved/7019>, it looks like:

* Chrome & Safari allow counters to increment on SVG internals, but never on `select` internals.
* Firefox & old-Edge allow counters to increment on SVG internals, and on *displayed* `option` elements, but *not* on `display:none` `option`s.

I think my preference right now is to say that SVG boxes can manipulate counters (those exist per spec, I'm pretty sure), but that other replaced-element internals cannot (unless otherwise defined).

**Comment by tabatkins** [comment]
Ugh, tho, <http://software.hixie.ch/utilities/js/live-dom-viewer/saved/7022> shows that Chrome does increment counters on `option` inside of `<select multiple>`.  That's terrible, this is terrible, the web is bad and a mistake.

**Comment by AmeliaBR** [comment]
So… I was thinking a little more about @kizu's comment about SVG, and wondered if it was specifically about non-rendered SVG elements that aren't technically `display: none`.  And what about `<use>` copies of SVG elements?

I made a slight variation: http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=7025

Chrome and Firefox both increment for non-rendered SVG elements, unless they explicitly have `display: none`.  Firefox gets completely lost with the `<use>` copy (the counter gets reset in the outer scope). Chrome just ignores it (not sure if that's normal for shadow DOM).  Old Edge reboots the tab helplessly at this monstrosity of a test; not sure which part is so painful (based on an [early version of the test](http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=7024), it had seemed to be incrementing for rendered, non-rendered, and shadow SVG content).

**Comment by MatsPalmgren** [comment]
> I think a strict reading of the spec says no - since `<select>` is replaced, [...]

It's wrong to say that `<select>` is a [replaced element](https://www.w3.org/TR/html5/rendering.html#replaced-elements), because it's not, it's a [widget](https://www.w3.org/TR/html5/rendering.html#widgets).
I would assume that `<option>/<optgroup>` are neither replaced elements, nor widgets, since they are not listed in those sections.

> ... its children don't generate CSS boxes

I tend to disagree with that. We certainly create boxes for `<option>/<optgroup>` internally in Gecko, albeit with severely restricted styling, and only for its text children etc. (Those limitations are completely artificial, btw. We could support pretty much all of CSS, and arbitrary child content, if the specs allowed it (in fact we did, but then we had to dumb it down to be compatible with other UAs/specs).)

> and on _displayed_ `option` elements, but not on `display:none` `option`s.

All UAs suppress incrementing counters on all `display: none` elements. I don't see why `<option>/<optgroup>` should be special in this respect.

> That's terrible, this is terrible, the web is bad and a mistake.

The web is terrible because the CSSWG put that in the spec :-p.

Seriously though, it's 2019 and I still can't do this:
```html
<select>
  <option><img src="Pineapple.png">Pineapple
  <option><img src="Apple.png">Apple
  <option><img src="Lemon.png">Lemon
</select>
```
Sigh. Now _that_ is truly terrible, because web authors routinely has to work around that in all sorts of terrible ways. It makes me sad because it would be so trivial to support if the specs just allowed it...

**Comment by MatsPalmgren** [comment]
BTW, this works perfectly fine in Gecko (meaning we create a grid container for it and render its child elements as grid items):
```html
<style>
option {
  display:grid;
  grid-template-columns: 100px auto;
}
</style>
<option><x>a</x><x>a</x>
```

Chrome also creates a box for it, but not a grid container(?), although its devtools still reports that the computed value is `grid-template-columns: 100px 1152px;` as if it were a grid container. Hmm.
Safari doesn't create a box at all.

**Comment by oliverdunk** [comment]
@MatsPalmgren Here's the layout tree in Chrome:

```
LayoutNGBlockFlow {HTML} at (0,0) size 800x600
    LayoutNGBlockFlow {BODY} at (8,8) size 784x584
      LayoutGrid {OPTION} at (0,0) size 784x20.19
        LayoutNGBlockFlow (anonymous) at (2,0) size 100x19.19
          LayoutText {#text} at (0,0) size 15x18
            text run at (0,0) width 15: "aa"
```

It looks like Chrome ignores any elements inside of the option element. I saw the same behaviour with divs, for example.

**Comment by oliverdunk** [comment]
What are the counter arguments to letting an option element be styled? It feels like if as @MatsPalmgren suggested, we allow images in them (which at a high level seems desirable), all CSS (and therefore counters) should work.

whatwg/html/issues/3596 seems relevant here.

**Comment by kizu** [comment]
Thinking about this, having counters on selects with `multiple` is probably the most “real” use case they could have — to show stuff like “N out of M options selected”. Counters on options of regular selects still could have viable use-cases (mostly for similar counting/validation help purposes), but as this is not something that was not available at all browsers, developers didn't research this enough.

Talking about if something like that should be present at all, I would confirm that, ideally, options _should_ generate boxes, and developers should have more control over what should be inside the options and how they should be displayed (also: how the option groups should be displayed and styled, but thats completely another issue). Actually, my experiments came from trying to find a way to at least display some extra images (flags for a country input, to be precise) for a selected value. Allowing images inside options would mean so much, and if select styling would be enhanced, I could live without counters on them (but then, if they would generate boxes, they would need to have counters, so… win/win).

**Comment by tabatkins** [comment]
The counter-argument to letting `option`s be styled is the same for all the "interesting" form controls: browsers have (and take advantage of!) significant leeway in displaying the elements. Load up a page with a `select` on Chrome desktop, Chrome android, and the Android Watch browser, for instance; in all three browsers there's significant variation in how the element and its contents are formatted and displayed. (And similar differences exist between all other browsers.)

We'd ideally like to figure out what to actually do about this, so we can define what properties actually work, but we've never had the time to do any of that & implementor interest to actually pursue any work. :(

**Comment by fantasai** [comment]
OK, so Tab and I are proposing to leave the case of descendants of replaced elements undefined for now.

**Comment by astearns** [comment]
@fantasai please make sure you have an edit for the note (I was surprised you closed this without that change)

**Comment by fantasai** [comment]
@astearns The note was added in 2988d2b9, forgot to link it to this issue.
